text/1003-cli-naming.rst

```diff
@@ -19,110 +19,238 @@ specified.
-subcommands:
+``configure`` uses ``<group> <command>`` whereas ``list-`` is flat
+------------------------------------------------------------------
+
+**Concern**: ``configure`` is inconsistent with ``list-``.
+
+**Decision**: It should be replaced by ``edgedb insert-config``,
+``edgedb set-config``, and``edgedb reset-config``.
+
+Why is ``server`` special?
+--------------------------
+
+**Concern**: the ``edgedb server`` CLI design outlined in RFC 1001
+directly conflicts with the ``<action>-<category>`` scheme.
+
+**Decision**: this sub-system is different because it is in fact
+a tool within a tool. Whereas other commands deal with a single
+database server, often a single database within that server, the
+``edgedb server`` sub-system deals with management of potentially
+many databases and server instances. With that in mind, keeping it
+separate makes sense.
+
+In particular, trying to shoehorn the sub-commands of ``edgedb server``
+into the ``<action>-<object>`` scheme would essentially move all of
+them into flat ``*-instance`` (or ``*-server``) sub-commands.
+
+Are ``list-*`` commands useful?
+-------------------------------
+
+**Concern**: there are nine ``list-*`` sub-commands which is over 36%
+of the available sub-commands. They mirror REPL functionality for listing
+items in the current database, for example ``\la`` or ``\list-aliases``
+will list available aliases. Are all of them necessary?
+
+**Decision**: the CLI should keep listings of entities that it can
+directly manipulate with other sub-commands. This includes:
+
+* ``list-databases`` (through ``create-database``);
+
+* ``list-ports`` (through ``configure``); and
+
+* ``list-roles`` (through ``alter-role``).
+
+Other listings should be removed as they are redundant with REPL
+functionality. Instead, the following form should be enabled::
+
+  $ echo "\list-object-types" | edgedb
+
+This is consistent with behavior of, say, the ``sqlite3`` CLI.
+
+The REPL help for ``\\`` commands is overloaded
+-----------------------------------------------
+
+**Concern**: It is convenient to expose the EdgeDB CLI commands directly in
+REPL. This can greatly simplify administrative tasks when the DB
+administrator has full DB admin rights but yet can't access the server shell.
+Sadly, the presence of those commands under the ``\\`` prefix mixes them
+up with the REPL-specific escape commands like ``\last-error``.
+
+**Decision**: Expose CLI-specific commands in the REPL via the ``!`` prefix,
+similar to how IPython exposes shell commands::
+
+  >>> \list-databases
+  tutorial
+
+  >>> SELECT 1;
+  {1}
+
+  >>> !dump db tutorial
+  done
+
+This way the internal REPL help system is not overloaded with rarely
+needed help on CLI commands and would only show the list of convenient ``\``
+commands with a hint that ``!help`` or ``!h`` can be used to list
+all CLI options.
+
+Name-spacing ``\help`` and ``!help`` is good for the usability, because
```

Comment:
We can probably have `\help`, which hides infrequent commands like `dump` and `\help --all`.